### PR TITLE
Adding in s3 HTTP challenge option

### PR DESCRIPTION
**Pull request by kingcdavid**
So in order for the s3 test to pass, we need to set up some env vars

TF_S3_BUCKET, this will need to be a real s3 bucket which you AWS creds have access to. if this is not set, it will skip the test
AWS_REGION, this will be the AWS region the s3 bucket is in

you will also have to have AWS credentials in your environment, this could be done in a number of ways:
https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/

### CODE REVIEW

**Comment by kingcdavid** [comment]
example of working test:

```console
go test -v -run TestAccACMECertificate_httpS3      
=== RUN   TestAccACMECertificate_httpS3
2023/08/01 12:07:37 [INFO] acme: Registering account for nobody@example.test
2023/08/01 12:07:37 [INFO] acme: Trying to resolve account by key
2023/08/01 12:07:37 [INFO] acme: Trying to resolve account by key
2023/08/01 12:07:37 [INFO] [test-s3.example.test, test-s32.example.test] acme: Obtaining bundled SAN certificate
2023/08/01 12:07:37 [INFO] [test-s3.example.test] AuthURL: https://localhost:14000/authZ/Ds0h3OSt2VYuJ6Y8ntyeHMShD1cT8ZSLIwp6BP6csXg
2023/08/01 12:07:37 [INFO] [test-s32.example.test] AuthURL: https://localhost:14000/authZ/yh1JYvTqHcZMGaBYiZt5i3_p8Vzy15gb-Tq8iymn6tw
2023/08/01 12:07:37 [INFO] [test-s3.example.test] acme: Could not find solver for: tls-alpn-01
2023/08/01 12:07:37 [INFO] [test-s3.example.test] acme: use http-01 solver
2023/08/01 12:07:37 [INFO] [test-s32.example.test] acme: Could not find solver for: tls-alpn-01
2023/08/01 12:07:37 [INFO] [test-s32.example.test] acme: use http-01 solver
2023/08/01 12:07:37 [INFO] [test-s3.example.test] acme: Trying to solve HTTP-01
2023/08/01 12:07:40 [INFO] [test-s3.example.test] The server validated our request
2023/08/01 12:07:40 [INFO] [test-s32.example.test] acme: Trying to solve HTTP-01
2023/08/01 12:07:44 [INFO] [test-s32.example.test] The server validated our request
2023/08/01 12:07:44 [INFO] [test-s3.example.test, test-s32.example.test] acme: Validations succeeded; requesting certificates
2023/08/01 12:07:44 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2023/08/01 12:07:44 [INFO] [test-s3.example.test] Server responded with a certificate.
2023/08/01 12:07:44 [INFO] acme: Trying to resolve account by key
2023/08/01 12:07:45 [INFO] acme: Trying to resolve account by key
2023/08/01 12:07:45 [INFO] acme: Trying to resolve account by key
2023/08/01 12:07:45 [INFO] acme: Deleting account for nobody@example.test
--- PASS: TestAccACMECertificate_httpS3 (9.27s)
PASS
ok  	github.com/vancluever/terraform-provider-acme/v2/acme	9.285s

```

**Comment by kingcdavid** [comment]
N.b This is pulling from acme-go@latest, so i guess we should wait till 4.14.0 is released?

File: acme/resource_acme_certificate_test.go [hunk 1]
```diff
@@ -37,7 +37,7 @@ func TestAccACMECertificate_basic(t *testing.T) {
 		ProviderFactories: testAccProviders,
 		ExternalProviders: testAccExternalProviders,
 		CheckDestroy:      testAccCheckACMECertificateStatus("acme_certificate.certificate", certificateStatusRevoked),
-		Steps: []resource.TestStep{
+			Steps: []resource.TestStep{
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:40]: @kingcdavid were these test files run though [`gofmt`](https://pkg.go.dev/cmd/gofmt)? This should likely be set up by your editor, but this indent looks out of place.

File: acme/resource_acme_certificate_test.go [hunk 2]
```diff
@@ -367,6 +367,38 @@ func TestAccACMECertificate_httpMemcache(t *testing.T) {
 	})
 }
 
+func TestAccACMECertificate_httpS3(t *testing.T) {
+	wantEnv := os.Environ()
+	s3_bucket := os.Getenv("TF_S3_BUCKET")
+	// If bucket not set, skip test
+	if s3_bucket == "" {
+		return
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:375]: [`t.Skip(...)`](https://pkg.go.dev/testing#T.Skip) should be used to skip tests.

File: acme/resource_acme_certificate_test.go [hunk 3]
```diff
@@ -652,6 +684,7 @@ func testAccCheckACMECertificateMemcacheTestServer() (func(), error) {
 	}, nil
 }
 
+
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:687]: Looks like a stray newline possibly due to a `gofmt` being necessary?

File: acme/resource_acme_certificate_test.go [hunk 4]
```diff
@@ -677,6 +710,30 @@ func testAccCheckACMECertificateProxyTestServer() (func(), error) {
...
+func testAccCheckACMECertificateS3ProxyTestServer(s3Bucket string, awsRegion string) (func(), error) {
+	target := fmt.Sprintf("%s.s3.%s.amazonaws.com",s3Bucket, awsRegion)
+
+	proxy := &httputil.ReverseProxy{
+        Director: func(req *http.Request) {
+            req.URL.Scheme = "https"
+            req.URL.Host = target
+            req.Host = target
+        },
+    }
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:722]: @kingcdavid looks like this definitely needs to be run through `gofmt`, as this brace is out of place as well.

**Review by vancluever — CHANGES_REQUESTED** [review]
@kingcdavid thanks for taking this on!

So looks like there's some things that need to be done for sure here, ahead of the next release of lego:

* Formatting: Can you run your changes through [gofmt](https://pkg.go.dev/cmd/gofmt).
* Documentation: Can you add an appropriate section under the [HTTP/TLS challenge section](https://github.com/vancluever/terraform-provider-acme/blob/main/docs/resources/certificate.md#using-http-and-tls-challenges) in the resource docs.
* Tests: I see that this might require an actual S3 bucket to run - did you happen to try getting it working with LocalStack so that it could be included in CI? The GH actions are not wired to real resources and I'm not planning on adding any.

Thanks again!

**Comment by kingcdavid** [comment]
Hey @vancluever

Yes sorry I should have formatted it, and updated the docs. I went with real s3 testing as openstack seemed quite heavy, but i can look at that. It mention in the openstack docs it significantly modifies a host, so may look to run it in docker if I can? What's your thoughts?

Thanks

**Comment by vancluever** [comment]
PS: it does look like [LocalStack has instructions for GH actions](https://docs.localstack.cloud/user-guide/ci/github-actions/), so it seems feasible we could use it.

You don't have to worry about that part though, so long as it works against a local endpoint that will be enough for me 🙂

**Comment by kingcdavid** [comment]
Ah sorry LocalStack does make much more sense

Though looking at the config, i'd have to make an upstream change to lego in order to point the AWS SDK to a local endpoint
i.e. https://github.com/localstack/localstack-aws-sdk-examples/blob/main/go/s3-basic-v2.go
which could get a little messy, unless you can think of a way to proxy it to local stack?

**Comment by vancluever** [comment]
Ah yeah, I see how the example uses the endpoint overrides in the SDK - we definitely don't want to use those. Looks like transparent endpoint injection is a pro feature too.

I think the fact that this is testable on S3 is good enough for now then! So long as the code remains relatively static and we can do ad hoc testing, I think that'll be fine. I might investigate other options just in case, but otherwise it should be okay.

File: acme/resource_acme_certificate_test.go [hunk 5]
```diff
@@ -367,6 +367,38 @@ func TestAccACMECertificate_httpMemcache(t *testing.T) {
 	})
 }
 
+func TestAccACMECertificate_httpS3(t *testing.T) {
+	wantEnv := os.Environ()
+	s3_bucket := os.Getenv("TF_S3_BUCKET")
+	// If bucket not set, skip test
+	if s3_bucket == "" {
+		t.Skip()
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:375]: _Note for @vancluever_ add a warning here mentioning that the test will be skipped due to missing env var

File: acme/resource_acme_certificate_test.go [hunk 6]
```diff
@@ -367,6 +367,38 @@ func TestAccACMECertificate_httpMemcache(t *testing.T) {
...
+	}
+	awsRegion := os.Getenv("AWS_REGION")
+	closeServer, err := testAccCheckACMECertificateS3ProxyTestServer(s3_bucket, awsRegion)
+	if err != nil {
+		panic(fmt.Errorf("TestAccACMECertificate_S3httpProxy: %s", err))
```
Comment (vancluever) [inline acme/resource_acme_certificate_test.go:380]: _Note for @vancluever_ fail test here instead of `panic()`

File: go.mod [hunk 7]
```diff
@@ -4,8 +4,9 @@ go 1.15
 
 require (
 	github.com/BurntSushi/toml v1.3.2
+	github.com/aws/aws-sdk-go v1.39.0 // indirect
 	github.com/davecgh/go-spew v1.1.1
-	github.com/go-acme/lego/v4 v4.13.1
+	github.com/go-acme/lego/v4 v4.13.4-0.20230727185640-f582d12f6580
```
Comment (vancluever) [inline go.mod:9]: _Note for @vancluever_ update/bump this to v4.14.0

**Review by vancluever — COMMENTED** [review]
Just documenting some necessary changes that I'm going to do to get this over the line, don't worry about doing anything here @kingcdavid :slightly_smiling_face:

**Review by vancluever — APPROVED** [review]
Merging this now! Thanks again for the contribution @kingcdavid!

**Pull request merged**